internal/mode/static/state/changed_predicate.go

```diff
@@ -1,37 +1,40 @@
 package state
 
-import "sigs.k8s.io/controller-runtime/pkg/client"
+import (
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
 
 // stateChangedPredicate determines whether upsert and delete events constitute a change in state.
 type stateChangedPredicate interface {
 	// upsert returns true if the newObject changes state.
-	upsert(oldObject, newObject client.Object) bool
+	upsert(oldObject, newObject client.Object, nsname types.NamespacedName) bool
 	// delete returns true if the deletion of the object changes state.
-	delete(object client.Object) bool
+	delete(object client.Object, nsname types.NamespacedName) bool
 }
 
 // funcPredicate applies the stateChanged function on upsert and delete. On upsert, the newObject is passed.
 // Implements stateChangedPredicate.
 type funcPredicate struct {
-	stateChanged func(object client.Object) bool
+	stateChanged func(object client.Object, nsname types.NamespacedName) bool
 }
 
-func (f funcPredicate) upsert(_, newObject client.Object) bool {
-	return f.stateChanged(newObject)
+func (f funcPredicate) upsert(_, newObject client.Object, nsname types.NamespacedName) bool {
+	return f.stateChanged(newObject, nsname)
```

Comment:
Events that led up to this change:

1. EndpointSlice resource is not tracked in store
2. `changeTrackingUpdater`'s `delete` function needs to be updated such that it can handle "deleting" a resource that is not tracked in the store. To do so we need to pass the `nsname` into the `stateChangedPredicate`'s `delete` function, then passed into `funcPredicate`'s `stateChanged` function.

Facing the question of what to do with the `upsert` function when calling `return f.stateChanged(newObject, what_to_put_here)` I had a couple of choices

1. Put an empty `types.NamespacedName{}`, keep the logic in `isReferenced` the same, and extract the `nsname` from the object when we're in `isReferenced`
2. Do something like `return f.stateChanged(newObject, types.NamespacedName{Namespace: newObject.GetNamespace(), Name: newObject.GetName()}
3. Pass `nsname` in as an argument. (Current changes)

I didn't do option 1 because it made `isReferenced` messy as it could either receive an empty nsname or a legit one through `delete`. I also didn't do option 2 because it made it so if `funcPredicate`'s `upsert` was called with a `nil` `newObject` (like it did in a couple tests), it would error/panic (I would just have to change a few tests to call `upsert` with a legit `newObject`). 

Option 3 however adds another parameter to `upsert` which is unneeded in the other predicates and moves the responsibility of passing in the `nsname` up to the caller of `upsert`. Any thoughts?

Reply:
@bjee19 for upsert you don't need to pass in the nsname, because you can get that name from the object. It's only necessary for the delete case because the object in that case is empty. You will just need to handle the nil case in the implementation of the stateChangedPredicate.

Reply:
Actually, there's no nil case for upsert. The new object will never be nil

Reply:
Okay! That was my original plan, but there are a couple of tests I would need to modify as they pass in `nil` for the `newObject` and I wasn't sure if we should require that `newObject` cannot be nil. (I'll probably add a panic statement there too then).